src/main.cpp

```diff
@@ -2472,6 +2472,10 @@ static DisconnectResult DisconnectBlock(const CBlock& block, CValidationState& s
         view.PopAnchor(SaplingMerkleTree::empty_root(), SAPLING);
     }
 
+    // History read/write is unconditional, since it is generated for entire
+    // chain regardless of the Heartwood upgrade activation height.
+    view.PopHistoryNode();
```

Comment:
Still todo: handle crossing a network upgrade activation boundary (where the tree switches to the prior epoch).